tests/simulator_test.cpp

```diff
@@ -2,82 +2,284 @@
+  } catch (const EXCEPTION_TYPE &e) {                                          \
+    EXPECT_STREQ(MESSAGE, e.what())                                            \
+        << " exception message is incorrect. Expected the following "          \
+           "message:\n\n"                                                      \
+        << MESSAGE << "\n";                                                    \
+  } catch (...) {                                                              \
+    FAIL() << "exception " << MESSAGE << " not thrown with expected type "     \
+           << #EXCEPTION_TYPE << "'!";                                         \
+  }
+
+using constants::kPi;
+using core::Ray;
+using core::RayHitData;
+using core::Vec3;
+
+const int kSkipNumCollectors = 37;
+const float kSkipFrequency = 1000;
+const Vec3 kVecZero(0, 0, 0);
+const Vec3 kVecUp(0, 0, 1);
+const Vec3 kVecX(1, 0, 0);
+const Vec3 kVecY(0, 1, 0);
+
+float deg2rad(float deg) { return 2 * kPi * deg / 360; }
+
+// TODO: add collision point check for every ray hit
+
+class FakeModel : public ModelInterface {
+
+public:
+  explicit FakeModel(bool empty) : empty_(empty){};
+  std::vector<objects::TriangleObj *> triangles() const override { return {}; }
+  float height() const override { return 0; }
+  float sideSize() const override { return 0; }
+  bool empty() const override { return empty_; }
+
+private:
+  bool empty_;
+};
+
+class EnergyCollectorTest : public ::testing::Test {
+protected:
+  // performs ray hit at at given energy collectors, |hitData| is modified to
+  // hold information where ray hit energyCollector. Returns true if hit
+  // occurred, false when there was no hit
+  [[nodiscard]] bool performHitCollector(
+      const std::vector<std::unique_ptr<objects::EnergyCollector>>
+          &energyCollectors,
+      const Ray &ray, RayHitData *hitData) {
+
+    for (const auto &collector : energyCollectors) {
+      if (collector->hitObject(ray, kSkipFrequency, hitData)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  void
+  printCollectors(const std::vector<std::unique_ptr<objects::EnergyCollector>>
+                      &energyCollectors) const {
+    for (const auto &collector : energyCollectors) {
+      std::cout << "Origin: " << collector->getOrigin() << "\n"
+                << "Radius: " << collector->getRadius() << "\n";
+    }
+  }
+
+  std::pair<objects::EnergyCollector, objects::EnergyCollector>
+  topCollectorsXAxis(
+      const std::vector<std::unique_ptr<objects::EnergyCollector>>
+          &energyCollectors) {
+    auto outputCollectors =
+        std::make_pair(objects::EnergyCollector(Vec3(0, 0, 0), 1),
+                       objects::EnergyCollector(Vec3(0, 0, 0), 1));
+    for (const auto &collector : energyCollectors) {
+      if (collector->getOrigin().z() >=
+              outputCollectors.first.getOrigin().z() &&
+          collector->getOrigin().x() ==
```

Comment:
Is this equivalent to just checking x() == 0?

Reply:
yep, fixed